README.md
# virtual-routing-demo
try push


src/history.txt

# 2018-05-13 10:08:57.509418
+-h text text

# 2018-05-13 10:09:52.349557
+--help

# 2018-05-13 10:10:16.120290
+asdf

# 2018-05-13 10:11:54.672743
+text sdf

# 2018-05-13 10:12:05.030432
+0v

# 2018-05-13 10:12:10.231623
+-v

# 2018-05-13 10:12:30.346497
+-v 1 2

# 2018-05-13 10:12:44.752379
+text start

# 2018-05-13 10:12:48.147369
+start

# 2018-05-13 10:38:19.869659
+-h

# 2018-05-13 10:46:37.811182
+add 1992 123.3123.321 34

# 2018-05-13 10:56:45.621727
+show interface

# 2018-05-13 10:57:31.026096
+send 7.7.1.1 7.7.1.2 asdf

# 2018-05-13 10:57:56.841552
+show route

# 2018-05-13 10:58:15.521138
+send 7.7.1.1 7.7.3.1 test

# 2018-05-13 10:58:55.044482
+add 7.7.3.0 24 7.7.1.2

# 2018-05-13 10:58:59.585625
+show route

# 2018-05-13 10:59:16.809064
+help

# 2018-05-13 10:59:23.124419
+help()

# 2018-05-13 10:59:39.831691
+send 7.7.1.1 7.7.3.1 test

# 2018-05-13 11:00:10.056930
+send 7.7.1.1 7.7.8.2 test

# 2018-05-13 11:00:30.521045
+add 7.7.8.0 24 7.7.1.2

# 2018-05-13 11:00:32.677562
+send 7.7.1.1 7.7.8.2 test

# 2018-05-13 11:00:55.550634
+add 7.7.8.0 24 7.7.3.1

# 2018-05-13 11:18:30.337610
+recv

# 2018-05-13 11:22:45.730034
+send 7.7.1.1 7.7.3.1 test

# 2018-05-13 11:22:56.073335
+send 7.7.1.1 7.7.1.2 test

# 2018-05-13 11:23:15.299836
+recv

# 2018-05-13 11:27:24.537167
+add 7.7.8.0 24 7.7.3.1

# 2018-05-13 11:28:12.760858
+show help

# 2018-05-13 13:14:58.677897
+show interface

# 2018-05-13 13:15:19.173630
+send 7.7.1.1 7.7.1.2 test

# 2018-05-13 13:15:42.353607
+recv

# 2018-05-13 15:11:34.086175
+send 7.7.1.1 7.7.0.2 test

# 2018-05-13 15:11:40.599809
+recv

# 2018-05-13 15:13:07.546989
+send 7.7.1.1 7.7.0.2 test!

# 2018-05-13 15:27:17.000549
+ 

# 2018-05-13 19:21:31.296367
+send 7.7.1.1 7.7.1.2 test

# 2018-05-13 19:21:33.864756
+recv

# 2018-05-13 19:25:05.656011
+q

# 2018-05-13 19:36:50.367766
+show route

# 2018-05-13 19:37:10.432945
+send 7.7.1.1 7.7.0.2 TEST!

# 2018-05-13 19:37:15.694264
+recv

# 2018-05-13 19:38:06.208481
+q

# 2018-05-13 19:38:27.823398
+q

# 2018-05-13 20:30:55.213113
+show route

# 2018-05-13 20:32:39.795918
+ss

# 2018-05-13 20:32:45.597679
+send 7.7.1.1 7.7.0.2 TEST!

# 2018-05-13 20:33:02.721615
+send 7.7.1.1 7.7.1.2 test

# 2018-05-13 20:33:09.650448
+show interface

# 2018-05-13 20:33:11.968950
+recv

# 2018-05-13 20:33:24.207427
+show route

# 2018-05-13 20:33:39.718854
+send 7.7.1.1 7.7.1.2 test

# 2018-05-13 20:33:47.288571
+recv

# 2018-05-13 20:34:54.183216
+q

# 2018-05-13 20:34:55.446535
+q

# 2018-05-13 20:34:57.733240
+q

# 2018-05-13 20:34:59.050564
+q

# 2018-05-13 20:35:00.328313
+q

# 2018-05-13 20:35:10.895809
+show route

# 2018-05-13 20:35:16.361207
+show interface

# 2018-05-13 20:35:21.188725
+send 7.7.1.1 7.7.1.2 test

# 2018-05-13 20:35:26.510672
+recv

# 2018-05-13 20:37:22.799963
+q

# 2018-05-13 20:37:33.672999
+send 7.7.1.1 7.7.1.2 test

# 2018-05-13 20:37:39.398891
+recv

# 2018-05-13 20:37:45.914730
+send 7.7.1.1 7.7.1.2 test

# 2018-05-13 20:38:56.000423
+send 7.7.1.1 7.7.0.2 TEST!

# 2018-05-13 20:39:01.489720
+recv

# 2018-05-13 20:39:52.261609
+send 7.7.0.2 7.7.1.1 TEST!

# 2018-05-13 20:40:01.681762
+recv

# 2018-05-13 20:40:36.239516
+send 7.7.0.2 7.7.6.2 TEST!

# 2018-05-13 20:40:52.100099
+show route

# 2018-05-13 20:41:04.315181
+show interface

# 2018-05-13 20:41:22.463276
+recv

# 2018-05-13 22:12:58.230557
+show route

# 2018-05-13 22:13:00.501122
+show help

# 2018-05-13 22:13:08.225941
+show route

# 2018-05-13 22:13:12.467920
+show tcp

# 2018-05-13 22:13:15.643675
+help

# 2018-05-13 22:13:17.449230
+show help

# 2018-05-13 22:13:31.795614
+add 7.7.8.0 24 7.7.3.1

# 2018-05-13 22:13:34.834859
+test send

# 2018-05-13 22:13:36.655200
+help

# 2018-05-13 22:13:38.102545
+show help

# 2018-05-13 22:13:53.040744
+send 1.1.1.1 'asfdasf'

# 2018-05-13 22:14:05.380228
+send 1.1.1.1 1.1.1.1 'asfdasf'

# 2018-05-13 23:07:50.012592
+test send

# 2018-05-13 23:08:47.239497
+show help

# 2018-05-13 23:09:15.612008
+send 8.8.2.2 8.8.4.2 test

# 2018-05-13 23:12:47.345143
+quit

# 2018-05-13 23:47:13.173971
+show route

# 2018-05-13 23:47:15.440572
+show tcp

# 2018-05-13 23:47:19.522973
+send 8.8.2.2 8.8.4.2 test

# 2018-05-13 23:47:21.849305
+send 8.8.2.2 8.8.4.2 testsafasdfasdf

# 2018-05-13 23:47:24.517767
+quit

# 2018-05-14 03:08:11.082902
+show route

# 2018-05-14 03:09:12.291874
+show help

# 2018-05-14 03:09:32.242023
+send 7.7.1.1 7.7.4.2 abc


src/console.py
"""
控制台
"""

import os
import json
import sys
import traceback
import time

from prompt_toolkit import prompt
from prompt_toolkit.history import FileHistory
from prompt_toolkit.auto_suggest import AutoSuggestFromHistory
from prompt_toolkit.contrib.completers import WordCompleter
from include.logger import logger



# NOTICE:如果需要使用终端，将以下代码复制到对应脚本的if __name__ == '__main__':下，并且import上面的库
# NOTICE:我是默认你们用终端之前已经import route并且初始化了网络层的，所以下面的代码直接使用网络层的接口
# TODO: from @yb, to @wyf. 我封装成了一个类。帮我check 一下有没有bug
class Console():
    def __init__(self, network_layer, route):
        self.network_layer = network_layer
        self.route = route
        self.Completer = WordCompleter(['show help', 'show tcp', 'show interface', 'show route', 'add', 'send', 'recv'],
                                    ignore_case=True)
        self.help_menu = [
            'show help\n\t: show the help message',
            'show tcp\n\t: show lower level tcp socket information',
            'show interface\n\t: show simulation interface status',
            'show route\n\t: show the route table',
            'send src_ip dest_ip data \n\t: send the data to a route\n\texample : send 8.8.1.1 8.8.4.2 teste!',
            'add dest_net net_mask final_ip \n\t: add an item in route table \n\texample : add 8.8.3.0 24 8.8.1.3 \n\tIt means that "to the net(8.8.3.0/24) via 8.8.4.2"',
            'delete dest_net net_mask\n\t: delete an item in route table \n\texample : delete 8.8.3.0 24"',
            'recv\n\t: no arguments',
        ]
    def task(self):
        while True:
            user_input = prompt('Route {}>'.format(self.network_layer.name), 
                                history=FileHistory('history.txt'),
                                auto_suggest=AutoSuggestFromHistory(),
                                completer=self.Completer,
                                )
            try:
                if user_input == '':
                    continue
                user_lines = user_input.split(';')
                for line in user_lines:
                    # 拆分用户参数
                    user_args = line.split()
                    main_action = user_args[0]

                    # 解析参数
                    if main_action == 'show':
                        if user_args[1] == 'interface':
                            self.route.link_layer.show_interface()
                        elif user_args[1] == 'tcp':
                            self.route.link_layer.show_tcp()
                        elif user_args[1] == 'route':
                            self.route.my_route_table.show()
                        elif user_args[1] == 'help':
                            print('This is help message!')
                            for help_msg in self.help_menu:
                                print('-'*40)
                                print(help_msg)
                    elif main_action == 'add':
                        # 往路由表中增加表项
                        self.route.my_route_table.update_item(user_args[1], int(user_args[2]), user_args[3])
                    elif main_action == 'delete':
                        # 删除某一项
                        self.route.my_route_table.delete_item(user_args[1], int(user_args[2]))
                    elif main_action == 'send':
                        # 发送信息
                        self.network_layer.send(user_args[1], user_args[2], user_args[3].encode('ascii'))
                    elif main_action == 'recv':
                        # 非阻塞接受IP包
                        ip_pkg = self.network_layer.recv()
                        if ip_pkg == None:
                            print('no receive!')
                        else:
                            print(ip_pkg)
                    elif main_action == 'debug':
                        if user_args[1] == 'start':
                            logger.disabled = False
                        if user_args[1] == 'stop':
                            logger.disabled = True
                    elif main_action == 'sleep':
                        time.sleep(int(user_args[1]))
                    elif main_action == 'suspend':
                        self.route.global_is_alive = 0
                        print('logout!  status:{}'.format(self.route.global_is_alive))
                    elif main_action == 'activate':
                        self.route.global_is_alive = 1
                        print('login!  status:{}'.format(self.route.global_is_alive))
                    elif main_action == 'quit':
                        os._exit(0)
            except IndexError:
                print('invalid command!')
                continue
            except Exception as e:
                # 捕获所有异常，并且打印错误信息
                print(traceback.format_exc())
                continue



if __name__ == "__main__":
    main()

def main():
    """ test """
    # 如果是作为模块被别的模块使用，就最好不要导入其他的名称空间了
    import route
    # 初始化网络层(内部还会初始化链路层)
    config_name = sys.argv[1]
    with open(config_name, 'r') as config_f:
        config = json.load(config_f)

    network_layer = route.NetworkLayer(config)
    console = Console(network_layer, route)
    console.task()